opensearch_py_ml/ml_models/metrics_correlation/wavelet_tools.py

```diff
@@ -0,0 +1,136 @@
+# SPDX-License-Identifier: Apache-2.0
+# The OpenSearch Contributors require contributions made to
+# this file be licensed under the Apache-2.0 license or a
+# compatible open source license.
+# Any modifications Copyright OpenSearch Contributors. See
+# GitHub history for details.
+
+"""wavelet_tools
+
+Tools for Haar discrete wavelet transform and signal reconstruction.
+
+Functions
+---------
+circularize
+    Circularizes a given filter to a specified length T.
+
+haar_matrix
+    Construct the DWT matrix for a Haar wavelet decomposition.
+
+haar_approx
+    Piecewise constant approximation of a signal by reconstruction
+    from a truncated Haar wavelet representation.
+"""
+
+import math
+
+# imports
+import torch
+import torch.nn.functional as F
+
+# hardcode haar wavelet and scale filters
+# HAAR_HI = torch.FloatTensor([-1/math.sqrt(2), 1/math.sqrt(2)])
+# HAAR_LO = torch.FloatTensor([1/math.sqrt(2), 1/math.sqrt(2)])
+
+
+def haar_approx(signal: torch.Tensor, truncate: int) -> torch.Tensor:
```

Comment:
Leaving in here as a pointer to future work: there are opportunities to further optimize these wavelet tools, which in turn might help make them more broadly usable. Ex: DFT vs wavelet matrix approach to computing the DWT, batch wavelet approximation of multiple series in parallel.